src/vaadin-overlay.html

```diff
@@ -599,15 +602,15 @@
           this._flushAnimation('closing');
         }
         this._attachOverlay();
+        if (!this.modeless) {
+          this._enterModalState();
+        }
         this.setAttribute('opening', '');
 
         const finishOpening = () => {
-          this.removeAttribute('opening');
           document.addEventListener('iron-overlay-canceled', this._boundIronOverlayCanceledListener);
 
-          if (!this.modeless) {
-            this._enterModalState();
-          }
+          this.removeAttribute('opening');
```

Comment:
Removing the `opening` attribute was moved to be the very last thing in `finishOpening()` so we can use it (e.g. in MutationObserver) to detect when the opening has been completely finished (including possible animations).

Other stuff (the modal state update) was moved out of this method to be run earlier/synchronously as it doesn't need to be tied to the completion of the possible animation.

Reply:
Similar changes to `finishClosing()` below.

Reply:
In my opinion, this refactor is important, in the future, we can use the opening/closing to check something because we know the status state: opening/ opened or closing/closed.

This fix addressed the bug and refactored code better. Thank you for the great fix.